### PR TITLE
JaegerExporter: Performance Improvements

### PR DESCRIPTION
**Pull request by CodeBlanch**
`JaegerUdpBatcher`, and the Thrift implementation it uses, is a total memory hog. When it is fed a span it allocates a buffer, writes the span into it, and then grabs the final size that will be needed off the buffer. It then throws away that buffer and puts the span into a list. During flush it goes through the list and re-writes the spans into another buffer. That buffer is then copied into yet another buffer on the UDP socket. Lots and lots of buffering going on. There's also over-use of Tasks. SendAsync on a socket really just writes into a buffer and returns synchronously.

So my theory going into this was...

1) We can reduce some work and memory usage by saving that initial buffer used to calculate size for the final transmission instead of saving the raw span and re-processing it.
2) We can reduce GC pressure by pooling buffers.
3) We can reduce allocations by using ValueTask where everything is actually running synchronously. Updated the ApacheThrift library for this part.

Benchmark:
```csharp
        [Benchmark]
        public async Task JaegerExporter_Batching()
        {
            using (var jaegerUdpBatcher = new JaegerUdpBatcher(
                new JaegerExporterOptions
                {
                    MaxPacketSize = int.MaxValue,
                    MaxFlushInterval = TimeSpan.FromHours(1)
                },
                new BlackHoleTransport()))
            {
                jaegerUdpBatcher.Process = new OpenTelemetry.Exporter.Jaeger.Implementation.Process("TestService", null);

                for (int i = 0; i < this.NumberOfBatches; i++)
                {
                    for (int c = 0; c < this.NumberOfSpans; c++)
                    {
                        await jaegerUdpBatcher.AppendAsync(this.testSpan, CancellationToken.None).ConfigureAwait(false);
                    }

                    await jaegerUdpBatcher.FlushAsync(CancellationToken.None).ConfigureAwait(false);
                }
            }
        }
```

Before:
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=3.1.200-preview-014977
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|                  Method | NumberOfBatches | NumberOfSpans |        Mean |     Error |    StdDev |        Gen 0 |      Gen 1 |     Gen 2 |   Allocated | Completed Work Items | Lock Contentions |
|------------------------ |---------------- |-------------- |------------:|----------:|----------:|-------------:|-----------:|----------:|------------:|---------------------:|-----------------:|
| **JaegerExporter_Batching** |               **1** |         **10000** |    **352.7 ms** |   **6.72 ms** |   **6.28 ms** |   **24000.0000** |  **6000.0000** |         **-** |   **193.35 MB** |               **3.0000** |                **-** |
| **JaegerExporter_Batching** |              **10** |         **10000** |  **3,530.4 ms** |  **56.53 ms** |  **52.88 ms** |  **242000.0000** | **27000.0000** | **1000.0000** |  **1933.44 MB** |               **2.0000** |                **-** |
| **JaegerExporter_Batching** |             **100** |         **10000** | **35,705.5 ms** | **654.64 ms** | **612.35 ms** | **2423000.0000** | **94000.0000** | **7000.0000** | **19334.34 MB** |               **5.0000** |                **-** |

After:
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=3.1.200-preview-014977
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|                  Method | NumberOfBatches | NumberOfSpans |        Mean |     Error |    StdDev |       Gen 0 |     Gen 1 | Gen 2 |  Allocated | Completed Work Items | Lock Contentions |
|------------------------ |---------------- |-------------- |------------:|----------:|----------:|------------:|----------:|------:|-----------:|---------------------:|-----------------:|
| **JaegerExporter_Batching** |               **1** |         **10000** |    **146.5 ms** |   **2.92 ms** |   **5.55 ms** |   **1000.0000** |         **-** |     **-** |   **10.94 MB** |               **1.0000** |                **-** |
| **JaegerExporter_Batching** |              **10** |         **10000** |  **1,421.6 ms** |  **19.13 ms** |  **16.95 ms** |  **13000.0000** |         **-** |     **-** |   **107.3 MB** |               **3.0000** |                **-** |
| **JaegerExporter_Batching** |             **100** |         **10000** | **14,385.2 ms** | **217.99 ms** | **203.91 ms** | **134000.0000** | **1000.0000** |     **-** | **1070.93 MB** |               **4.0000** |                **-** |

Basically it is faster and uses less memory.

### CODE REVIEW

File: lib/Thrift/Transports/Client/TBufferedClientTransport.cs [hunk 1]
```diff
@@ -87,7 +94,11 @@ public override void Close()
                 _inputBuffer.Capacity = _bufSize;
             }
 
+#if NETSTANDARD2_1
+            var got = await _inputBuffer.ReadAsync(new Memory<byte>(buffer, offset, length), cancellationToken).ConfigureAwait(false);
```
Comment (MikeGoldsmith) [inline lib/Thrift/Transports/Client/TBufferedClientTransport.cs]: Instead of allocating a new `Memory<byte>`, you could use `MemoryPool<byte>.Shared.Rent()` instead. Check out it's suggested usage [here](https://docs.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines#memoryt-and-the-ownerconsumer-model).

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 namespace OpenTelemetry.Exporter.Jaeger.Implementation
 {
-    public interface IJaegerUdpClient : IDisposable
+    public interface IJaegerClient : IDisposable
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs]: Do we need to rename the interface? It would be a break change to anyone who's used the public interface.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs [hunk 3]
```diff
@@ -43,7 +43,7 @@ public static class JaegerConversionExtensions
         private const long TicksPerMicrosecond = TimeSpan.TicksPerMillisecond / 1000;
         private const long UnixEpochMicroseconds = UnixEpochTicks / TicksPerMicrosecond; // 62,135,596,800,000,000
 
-        public static JaegerSpan ToJaegerSpan(this SpanData span)
+        public static void ToJaegerSpan(this SpanData span, out JaegerSpan jaegerSpan)
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs]: Why change the signature to use out param instead of returning JaegerSpan?

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs]: No reason, really. Overstepping my bounds ;) It just seemed to add more value with this name. It's not representing anything UDP specific, just any old client that can be plugged into `JaegerThriftClientTransport` (so long as it can satisfy the signature). Seemed like if you were going to switch out the client it would because you wanted to change the protocol, not re-implement UDP. But happy to change it back, just confirm for me please. Still alpha so not sure if there is an opportunity to be breaking.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerConversionExtensions.cs]: Just one of the things I tried (of many) to get the allocations down. I'll put it back, just tested and this had no impact on the perf. I made some things structs which were only being used while the logic was executing to build up the buffer, I thought that would reduce allocations but the memory actually jumped up! After a lot of messing around, it seemed to be this pattern:
```csharp
if (this.References is IEnumerable<JaegerSpanRef> references)
```
Seems like that "is" operation does a boxing and actually copies the data from the stack to the heap? That was a first to see, to be honest, and I'm not totally sure what was going on. When I switched them to "!= null" checks, memory started behaving as I was expecting. I haven't had a chance to dig more into it.

File: lib/Thrift/Transports/Client/TBufferedClientTransport.cs (hunk 1 above)
Comment (CodeBlanch) [reply lib/Thrift/Transports/Client/TBufferedClientTransport.cs]: Sorry about all the changes! I thought it would be quick and easy... a couple of tweaks here and there... and it just kept unraveling more and more on me. Taking a look at MemoryPool stuff. I'll reply back.

**Comment by SergeyKanzhelev** [comment]
@CodeBlanch while you are on it - do you think it may be a good idea to build Thrift into the exporter code instead of a separate NuGet? I never published it on nuget.org. Mostly because doesn't want confusion.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs (hunk 2 above)
Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs]: @MikeGoldsmith I don't think it's critical to keep names if it's reasonable rename. Have you seen anybody taking dependency on it already?

File: lib/Thrift/Transports/Client/TBufferedClientTransport.cs (hunk 1 above)
Comment (CodeBlanch) [reply lib/Thrift/Transports/Client/TBufferedClientTransport.cs]: @MikeGoldsmith I tried to make it work with `ReadOnlyMemory<byte>` being passed in on writes and `Memory<byte>` being passed in on reads but it started to get really nasty. Stream doesn't have...
```csharp
ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default);
```
...until .NET Standard 2.1 which means to support net46 & standard 2.0 we have to end up with ([], offset, count) pattern in order to write to the UDP socket, which doesn't seem possible from the new memory types?

**Comment by CodeBlanch** [comment]
@SergeyKanzhelev It definitely crossed my mind when I was in there updating all the files. There's a whole bunch of stuff in the Thirft library we're not using. None of the server stuff, none of the reading stuff, none of the transports (we have our own transport in the JaegerExporter). Could probably shed a lot of it if we pulled in just what is needed. We use the TCompactProtocol and the entities, basically. We're really just building a buffer for the batch and then dropping that on a UDP socket (buffer). A lot (all?) of the asynchronicity in Thirft is just overhead for us because we're not actually doing any IO.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs (hunk 2 above)
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Exporter.Jaeger/Implementation/IJaegerClient.cs]: Yeah, I don't think there is anyone who depends on it. I think I queried it just because it was mixed in with the other changes.

I'm not against it and think the updated name is better / clearer.

**Comment by SergeyKanzhelev** [comment]
@CodeBlanch Thrift used here is basically a copy of https://github.com/jaegertracing/jaeger-client-csharp/tree/master/src/Thrift/netcore/Thrift. Since we are changing our version I'd suggest we just compile it into the exporter code. Maybe as a separate PR

@yurishkuro do you have more context and/or suggestions here?

**Comment by yurishkuro** [comment]
cc @Falco20019 re thrift/nuget question

Re the top-level performance concern - it sounds like the implementation is suboptimal. In some other Jaeger clients we usually use just two buffers, one for each span to get the encoded size, and the final buffer with the combined payload. By going a bit lower-level in Thrift encoding it's possible to use just a single buffer where the spans are appended in the final encoding.

Is it necessary to support Jaeger format directly? There's always the path of otel-sdk -> otel-collector -> jaeger.

**Comment by CodeBlanch** [comment]
Random question, IJaegerUpdBatcher has this interface:
```csharp
    public interface IJaegerUdpBatcher : IDisposable
    {
        Process Process { get; }

        Task<int> AppendAsync(JaegerSpan span, CancellationToken cancellationToken);

        Task<int> CloseAsync(CancellationToken cancellationToken);

        Task<int> FlushAsync(CancellationToken cancellationToken);
    }
```
CloseAsync just calls FlushAsync and isn't called ever. Should we drop it?

**Comment by Falco20019** [comment]
@CodeBlanch CloseAsync should be called in Dispose to ensure the spans are send to the agent even if not closed manually before. Otherwise they might get lost if not yet flushed. That’s at least how it‘s implemented in Jaeger. Haven’t looked at the implementation here so far.

@yurishkuro Have to check more into your question once in the office.

**Comment by CodeBlanch** [comment]
**Updates...**

1. Updated JaegerUdpBatcher to call CloseAsync during Dispose based on what @Falco20019 said. Tested, turns out Dispose is never called on JaegerUdpBatcher because BatchingSpanProcessor and SimpleSpanProcessor don't dispose of their Exporter when they are disposed. Fixed that up.

2. Switched to the official [ApacheThrift](https://www.nuget.org/packages/ApacheThrift) package. The performance looked good so I removed the fork. The package is a bit tricky because it bundles a net35 & net45 binary which has a different API. I had to do some kungfu in the csproj so we only use the netstandard2.0 binary.

**Work remaining...**

I have to merge in #486 which is going to be a bit of effort. I'll work on that this weekend.

**Comment by CodeBlanch** [comment]
Update...

* #486 merged into this PR.
* Switched ApacheThrift reference to PrivateAssets="All".
* Updated the description & performance numbers on the PR description to reflect latest master vs this branch.

**Comment by SergeyKanzhelev** [comment]
so cool! Thank you

**Comment by CodeBlanch** [comment]
@SergeyKanzhelev My goal is for the JaegerExporter to allocate enough memory for 1 batch and then re-use that for all subsequent batches it sends. Still messing with it, I'll open a new PR if I make significant headway.
